Add to the Product Image block the support for the Product Query block

gigitux opened the pull request:
This PR adds to the Product Image block the support for the Product Query block.
Now, on the client side, when the Product Image block is used in the Product Query block, the markup will be server side generated. Also, no javascript related to the Product Image Block will be loaded. 


 It has been necessary:
- refactor `withProductDataContext HOC` to use a store. This was very important because every time in the Product Query, the Post Template block is clicked, the component recreates itself, so it is necessary to cache the result to avoid flickering and useless fetch requests. 
- remove the possibility of using the Product Image block alone. Refactoring `withProductSelector HOC`  was too much time-consuming. 
- Hide the Image Sizing option (that contains a link for the Customizer) on the Product Image block when a block theme is active. This is necessary because Customizer will not be available on block themes anymore https://github.com/woocommerce/woocommerce/pull/34022


Fixes #6785

<!-- Don't forget to update the title with something descriptive. -->
<!-- If you can, add the appropriate labels -->


### Testing

N/A

#### User Facing Testing

<!-- Write these steps from the perspective of a "user" (merchant) familiar with WooCommerce. No need to spell out the steps for common setup scenarios (eg. "Create a product"), but do be specific about the thing being tested. Include screenshots demonstrating expectations where that will be helpful. -->

1. Create a post/page and add the All Products block. Be sure that there isn't any regression on the editor side and frontend side for all Atomic blocks (all the atomic blocks are using `withProductDataContext HOC`.
2. Create a post/page and add the Product Query block. Be sure that all the options related to the block work (eg: "show on sale badge"). Click on different Product Image block and be sure that there isn't any flickering effect (example attached).
3. Save it.
4. Open the frontend side and be sure that:
- no JS related to the Product Image block is loaded (the block is rendered on the server).
- The UI of the block is identical to on the editor side.
- The settings are reflected correctly (eg: "show on sale badge" disabled -> no on sale badge should be visible)


![Screen Capture on 2022-08-09 at 14-45-31 (2)](https://user-images.githubusercontent.com/4463174/185429411-e00f4628-c1ef-4405-8f02-c95847209fb1.gif) 
Flickering effect before refactoring the `withProductDataContext HOC`.


* [x] Do not include in the Testing Notes <!-- Check this box if this PR can't be tested by users (ie: it doesn't include user-facing changes or it can't be tested without manually modifying the code). -->

### WooCommerce Visibility

<!-- Check this [this doc](../docs/blocks/feature-flags-and-experimental-interfaces.md) to see if the change is visible in WC core or not (part of the feature plugin or experimental)-->

* [ ] WooCommerce Core
* [ ] Feature plugin
* [x] Experimental

gigitux committed:
Adds to the Product Image Block the support for the Product Query Block

Adds to the Product Image Block the support for the Product Query Block #6911

gigitux committed:
Add to the Product Image Block the support for the Product Query Block

dinhtungdu commented on src/BlockTypes/ProductImage.php:
```diff
@@ -48,12 +50,156 @@ protected function get_block_type_supports() {
+		 * @param array $attributes  Any attributes that currently are available from the block.
+		 *                           Note, this will be empty in the editor context when the block is
+		 *                           not in the post content on editor load.
+		 */
```
This docblock is misaligned.

dinhtungdu commented on assets/js/shared/hocs/with-product-data-context.js:
```diff
@@ -1,54 +1,60 @@
+	const productFromAPI = useSelect( ( select ) => {
+		if ( id > 0 ) {
+			const prod = select( STORE_NAME ).getProduct( id );
+			const hasFinished = select( STORE_NAME ).hasFinishedResolution(
+				'getProduct',
+				[ id ]
+			);
+
+			return {
+				product: prod || null,
+				isLoading: ! hasFinished,
+			};
 		}
-	}, [ productId ] );
 
-	if ( ! isLoading && ! product ) {
-		return null;
+		return {
+			product: null,
+			isLoading: false,
+		};
+	} );
```
Can we use our existing hook here to avoid creating another store? You mentioned we should use `@woocommerce/data`, so I think eventually, we will delete the custom store introduced in this PR. So I think using the existing hook can make it easier to migrate to `@woocommerce/data` later.

I tried to use `useStoreProducts` instead and get the same result: https://github.com/woocommerce/woocommerce-blocks/compare/add/product-image-product-grid...try/use-existing-hook-for-hoc

dinhtungdu commented on assets/js/blocks/product-query/variations/products-on-sale.tsx:
```diff
@@ -43,7 +43,7 @@ if ( isExperimentalBuild() ) {
 				'core/post-template',
 				{},
-				[ [ 'core/post-title' ], [ 'core/post-featured-image' ] ],
+				[ [ 'core/post-title' ], [ 'woocommerce/product-image' ] ],
```
Same for the on sale variation.
```suggestion
				[ [ 'woocommerce/product-image' ], [ 'core/post-title' ] ],
```

dinhtungdu commented on assets/js/blocks/product-query/variations/product-query.tsx:
```diff
@@ -40,7 +40,7 @@ if ( isExperimentalBuild() ) {
 				'core/post-template',
 				{},
-				[ [ 'core/post-title' ], [ 'core/post-featured-image' ] ],
+				[ [ 'core/post-title' ], [ 'woocommerce/product-image' ] ],
```
For the product grid, the image is usually to put above the title, so I think it makes sense to have it default here.
```suggestion
				[ [ 'woocommerce/product-image' ], [ 'core/post-title' ] ],
```

dinhtungdu reviewed (commented):
@gigitux This is great! The product query works flawlessly in my test. 🎉 

Besides review comments, I have a question related to the block rename. Can you explain further the reason why we need to rename the `image` component to `product-image`?

gigitux committed:
address feedback

gigitux committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce-blocks into add/product-image-product-grid

gigitux commented:
Thanks for the review! I addressed your feedback!
> Besides review comments, I have a question related to the block rename. Can you explain further the reason why we need to rename the `image` component to `product-image`?

Good point. I did some experiments because the idea was to migrate to `block.json`, but it wasn't possible (internal discussion -> p1660815701384609-slack-C02UBB1EPEF). I reverted the changes related to the block rename!

dinhtungdu reviewed (approved):
Thanks for the updates! The default layout looks much better now. I just have some minor suggestions:
- Can we change the default view to the Grid view?
- The post title is using `h2` and it's too big, can we change it to `h3` and change the size to `L`.

[Question] We have the product element title block, do we plan to make it compatible with the product query block?

I will be AFK until next Tuesday, so I'm preapproving this PR so I won't block you from merging it. Please request reviews from other team members if needed!

gigitux committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce-blocks into add/product-image-product-grid

gigitux committed:
set grid view and font-size L as default

gigitux commented:
> Thanks for the updates! The default layout looks much better now. I just have some minor suggestions:
> 
> * Can we change the default view to the Grid view?
> * The post title is using `h2` and it's too big, can we change it to `h3` and change the size to `L`.
> 
Thanks for the suggestion. I addressed your feedback, this is how it looks like the block now:

![image](https://user-images.githubusercontent.com/4463174/187698743-ad6ad1d7-eab2-475c-ac32-47cb4ef4d603.png)


> [Question] We have the product element title block, do we plan to make it compatible with the product query block?
> 

I'm not sure yet. I'm inclined to use the `Post Title` block if it doesn't introduce any regression.

dinhtungdu reviewed (approved):
@gigitux thank you so much! The update is LGTM! 🎉

opr commented on src/BlockTypes/AbstractBlock.php:
```diff
@@ -76,14 +76,16 @@ public function __construct( AssetApi $asset_api, AssetDataRegistry $asset_data_
 	 * @return string Rendered block type output.
 	 */
-	public function render_callback( $attributes = [], $content = '' ) {
+	public function render_callback( $attributes = [], $content = '', $block ) {
```
On PHP 8 this line is giving me the following error:

`( ! ) Deprecated: Required parameter $block follows optional parameter $attributes in [snip]/app/public/wp-content/plugins/woocommerce-blocks/src/BlockTypes/AbstractBlock.php on line 82`
